godot-macros/src/docs/markdown_converter.rs

```diff
@@ -5,91 +5,330 @@
  */
 
-//! Converts [Markdown](https://en.wikipedia.org/wiki/Markdown) to [BBCode](https://en.wikipedia.org/wiki/BBCode).
+//! Converts [Markdown](https://en.wikipedia.org/wiki/Markdown) to Godot-compatible [BBCode](https://en.wikipedia.org/wiki/BBCode).
 
 use markdown::mdast as md;
 use markdown::{to_mdast, ParseOptions};
-use std::collections::HashMap;
+use std::collections::{BTreeMap, HashMap};
 
-pub fn to_bbcode(md: &str) -> String {
+/// Converts the provided Markdown string to BBCode suitable for Godot's docs renderer.
+/// Simulates any missing features (e.g. tables) with a best-effort approach.
+pub fn to_bbcode(md_text: &str) -> String {
     // to_mdast() never errors with normal Markdown, so unwrap is safe.
-    let n = to_mdast(md, &ParseOptions::gfm()).unwrap();
+    let root = to_mdast(md_text, &ParseOptions::gfm()).unwrap();
 
-    let definitions = n
+    // Collect link/image definitions (for reference-style links).
+    let definitions = root
         .children()
-        .unwrap() // root node always has children
+        .unwrap() // root always has children
         .iter()
-        .filter_map(|n| match n {
+        .filter_map(|node| match node {
             md::Node::Definition(def) => Some((&*def.identifier, &*def.url)),
             _ => None,
         })
         .collect::<HashMap<_, _>>();
 
-    walk_node(&n, &definitions).unwrap_or_default()
-}
+    // Convert the root node to BBCode.
+    let mut converter = BBCodeConverter::new(&definitions);
+    let content = converter.walk_node(&root, 0).unwrap_or_default();
 
-fn walk_node(node: &md::Node, definitions: &HashMap<&str, &str>) -> Option<String> {
-    use md::Node::*;
+    // Append footnotes at the bottom if any.
+    if !converter.footnote_defs.is_empty() {
+        let notes = converter
+            .footnote_defs
+            .iter()
+            .map(|(idx, text)| format!("{} {}", BBCodeConverter::superscript(*idx), text))
+            .collect::<Vec<_>>()
+            .join("[br]");
+        format!("{content}[br][br]{notes}")
+    } else {
+        content
+    }
+}
 
-    let bbcode = match node {
-        Root(root) => walk_nodes(&root.children, definitions, "[br][br]"),
+/// Manages the context needed to convert Markdown AST to Godot-compatible BBCode.
+pub struct BBCodeConverter<'a> {
+    /// Link/image definitions from the Markdown AST.
+    definitions: &'a HashMap<&'a str, &'a str>,
 
-        InlineCode(md::InlineCode { value, .. }) => format!("[code]{value}[/code]"),
+    /// Footnote label -> numeric index.
+    footnote_map: HashMap<String, usize>,
 
-        Delete(delete) => format!("[s]{}[/s]", walk_nodes(&delete.children, definitions, "")),
+    /// Footnotes (index -> rendered text), sorted by index.
+    pub footnote_defs: BTreeMap<usize, String>,
 
-        Emphasis(emphasis) => format!("[i]{}[/i]", walk_nodes(&emphasis.children, definitions, "")),
+    /// Next footnote index to assign.
+    current_footnote_index: usize,
+}
```

Comment:
Why is one field public, but not the others? If we need read access outside, should we maybe provide a getter with the most limited API possible?

Reply:
Off the top of my head I don't think there was a reason, probably just snuck in an extra `pub`, but I'm not looking at the code right now. I'll take a look when I free up.

Reply:
Checked, it wasn't necessary. Removed.